Fixed #34262 -- Added support for AnyValue for MySQL, Oracle, and Postgresql 16+. (#19196)
django/django: docs/topics/db/aggregation.txt

    >>> Book.objects.annotate(num_authors=Count("authors")).aggregate(Avg("num_authors"))
    {'num_authors__avg': 1.66}

> When you are generating an aggregate by using ``.values()`` clause to create the

Author: I gave this a sub heading and linked to it in the AnyValue admonition. I wasn't entirely sure of the wording though. Does it seem all right?

> When you are generating an aggregate by using ``.values()`` clause to create the

Reviewer: Yes :star:

    >>> Book.objects.annotate(num_authors=Count("authors")).aggregate(Avg("num_authors"))
    {'num_authors__avg': 1.66}

When you are generating an aggregate by using ``.values()`` clause to create the
grouping for the annotation on MySQL, and the ``ONLY_FULL_GROUP_BY`` SQL mode
is enabled, it may be necessary to apply :class:`~django.db.models.AnyValue` if

> the annotation includes a mix of aggregated and non-aggregated functions.

Reviewer 2: I think we usually use `aggregate` and not `aggregated` to refer to such functions.

```suggestion
the annotation includes a mix of aggregate and non-aggregate expressions.
```

Take the following example:

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest
    >>> Book.objects.values(greatest_pages=Greatest("pages", 600)).annotate(
    ...     num_authors=Count("authors"),
    ...     pages_per_author=F("greatest_pages") / F("num_authors"),
    ... ).aggregate(Avg("pages_per_author"))

This creates groups of books based on the SQL column ``GREATEST(pages, 600)``.
One unique group consists of books with 600 pages or less, and other unique

> groups will consist of books with the same pages. The ``pages_per_author``
> involves a mix of aggregated and non-aggregated functions, because
> ``greatest_page`` is a non-aggregated function, whereas the ``num_authors`` is
> an aggregated function.

Reviewer 2: ```suggestion
groups will consist of books with the same pages. The ``pages_per_author``
annotation is composed of aggregate and non-aggregate expressions,
``num_authors`` is an aggregate expression while ``greatest_page`` isn't.
```

The column in the group by clause is the expression for ``greatest_pages``.
MySQL is not able to determine that the ``greatest_pages`` referenced in
``pages_per_author`` is functionally dependent on the column in the group by

> clause, and it will raise the
> ``OperationalError: incompatible with sql_mode=only_full_group_by`` error.

Reviewer 2: I wonder if we should use _might_ here as MySQL might get smarter about that in the future

```suggestion
clause, and it might raise a
``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
```

> To avoid the ``OperationalError``, you can wrap the non-aggregated function

Reviewer 2: ```suggestion
To avoid the ``OperationalError``, you can wrap the non-aggregate expression
```

with :class:`~django.db.models.AnyValue`.

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600),
    ... ).annotate(
    ...     num_authors=Count("authors"),
    ...     pages_per_author=AnyValue(F("greatest_pages")) / F("num_authors"),
    ... ).aggregate(Avg("pages_per_author"))
    {'pages_per_author__avg': 532.57143333}

The other supported databases will not encounter the ``OperationalError`` in the
example above, because they will detect the functional dependency. In general,
``AnyValue`` is useful for working around columns in the select list that
involve non-aggregated functions or complex expressions that are not determined
by the database to be functionally dependent on the columns in the grouping clause.

> It is an optimized alternative to the ``Min`` and ``Max`` aggregate functions.

Reviewer 2: Not sure if this part is relevant as `ANY_VALUE` is allowed to pick any value even if `ORDER BY` is specified AFAIK. In other words, the fact some backends will honor the `ORDER BY` clause and pick the first item is an implementation detail?

Reviewer 2: Sorry for the ambiguity, I meant to target only the line about the optimization.

    >>> Book.objects.annotate(num_authors=Count("authors")).aggregate(Avg("num_authors"))
    {'num_authors__avg': 1.66}

When you are generating an aggregate by using ``.values()`` clause to create the grouping
for the annotation, it may be necessary to apply :class:`~django.db.models.AnyValue`

> when the annotation includes a mix of aggregated and non-aggregated functions. This

Reviewer 2: Are we aware that this might be a problem on other backends than MySQL?

Author: It doesn't seem to be a problem for other backends as far as I can tell when removing AnyValue from `test_any_value_aggregate_clause`. I've updated this to make it clear that this is a problem for MySQL.

> happens when the database is not be able to determine the functional dependency of the
> columns in the non-aggregated functions to the grouping column.

Reviewer 2: ```suggestion
happens when the database is not able to determine the functional dependency of the
columns in the non-aggregated functions to the grouping column.
```

Take the following example:

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest

>     >>> Book.objects.annotate(greatest_pages=Greatest("pages", 600)).values(
>     ...     "greatest_pages"
>     ... ).annotate(

Reviewer 2: ```suggestion
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600)
    ... ).annotate(
```

    ...     num_authors=Count("authors"),
    ...     pages_per_author=F("greatest_pages") / F("num_authors"),
    ... ).aggregate(
    ...     Avg("pages_per_author")
    ... )

This creates groups of books based on the SQL column ``GREATEST(pages, 600)``.
One unique group consists of books with 600 pages or less, and other unique
groups will consist of books with the same pages. The ``pages_per_author``
involves a mix of aggregated and non-aggregated functions, because
``greatest_page`` is a non-aggregated function, whereas the ``num_authors`` is
an aggregated function.

The database is not able to determine that the ``greatest_pages`` used in
``pages_per_author`` is functionally dependent on the column in the group by
clause. In MySQL, this will happen if the ``sql_mode`` contains
``ONLY_FULL_GROUP_BY``, and it will raise the

> ``OperationalError: incompatible with sql_mode=only_full_group_by`` error.

Reviewer 3: Reading this as someone who doesn't know much about the feature, do I need to be worried about this error with non-mysql databases? If the answer is no, do we still want to mention the use case of why we should use `AnyValue` with non-mysql databases?

Reviewer 2: > Reading this as someone who doesn't know much about the feature, do I need to be worried about this error with non-mysql databases?

I think we don't and share your sentiment that the documentation should be adjusted accordingly.

To avoid the ``OperationalError``, you can wrap the non-aggregated function
with :class:`~django.db.models.AnyValue`.

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest

>     >>> Book.objects.annotate(greatest_pages=Greatest("pages", 600)).values(
>     ...     "greatest_pages"
>     ... ).annotate(

Reviewer 2: ```suggestion
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600),
    ... ).annotate(
```

    >>> Book.objects.annotate(num_authors=Count("authors")).aggregate(Avg("num_authors"))
    {'num_authors__avg': 1.66}

.. _aggregation-mysql-only-full-group-by:

> Aggregating with MySQL ONLY_FULL_GROUP_BY enabled
> -------------------------------------------------

Reviewer: ```suggestion
Aggregating with MySQL ``ONLY_FULL_GROUP_BY`` enabled
-----------------------------------------------------
```

I am also tempted to move this section to the bottom of the page as it's quite niche

> When you are generating an aggregate by using ``.values()`` clause to create the
> grouping for the annotation on MySQL, and the ``ONLY_FULL_GROUP_BY`` SQL mode
> is enabled, it may be necessary to apply :class:`~django.db.models.AnyValue` if
> the annotation includes a mix of aggregate and non-aggregate expressions.

Reviewer: ```suggestion
When using the ``values()`` clause to group query results for annotations in
MySQL with the ``ONLY_FULL_GROUP_BY`` SQL mode enabled, you may need to apply
:class:`~django.db.models.AnyValue` if the annotation includes a mix of
aggregate and non-aggregate expressions.
```

(with each of these suggestions, only take bits you like)

Take the following example:

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest
    >>> Book.objects.values(greatest_pages=Greatest("pages", 600)).annotate(
    ...     num_authors=Count("authors"),
    ...     pages_per_author=F("greatest_pages") / F("num_authors"),
    ... ).aggregate(Avg("pages_per_author"))

This creates groups of books based on the SQL column ``GREATEST(pages, 600)``.
One unique group consists of books with 600 pages or less, and other unique
groups will consist of books with the same pages. The ``pages_per_author``
annotation is composed of aggregate and non-aggregate expressions,
``num_authors`` is an aggregate expression while ``greatest_page`` isn't.

> The column in the group by clause is the expression for ``greatest_pages``.
> MySQL is not able to determine that the ``greatest_pages`` referenced in
> ``pages_per_author`` is functionally dependent on the column in the group by
> clause, and it might raise a
> ``OperationalError: incompatible with sql_mode=only_full_group_by`` error.
> To avoid the ``OperationalError``, you can wrap the non-aggregate expression
> with :class:`~django.db.models.AnyValue`.

Reviewer: ```suggestion
Since the grouping is based on the ``greatest_pages`` expression, MySQL may be
unable to determine that ``greatest_pages`` (used in the ``pages_per_author``
expression) is functionally dependent on the grouped column. As a result, it
may raise an error like:

.. code-block:: pytb

    OperationalError: incompatible with sql_mode=only_full_group_by

To avoid this, you can wrap the non-aggregate expression with
:class:`~django.db.models.AnyValue`.
```

.. code-block:: pycon

    >>> from django.db.models import F, Count, Greatest
    >>> Book.objects.values(
    ...     greatest_pages=Greatest("pages", 600),
    ... ).annotate(
    ...     num_authors=Count("authors"),
    ...     pages_per_author=AnyValue(F("greatest_pages")) / F("num_authors"),
    ... ).aggregate(Avg("pages_per_author"))
    {'pages_per_author__avg': 532.57143333}

> The other supported databases will not encounter the ``OperationalError`` in the
> example above, because they will detect the functional dependency. In general,
> ``AnyValue`` is useful for working around columns in the select list that
> involve non-aggregate functions or complex expressions that are not determined
> by the database to be functionally dependent on the columns in the grouping clause.

Reviewer: ```suggestion
Other supported databases do not encounter the ``OperationalError`` in the
example above because they can detect the functional dependency. In general,
``AnyValue`` is useful when dealing with select list columns that involve
non-aggregate functions or complex expressions not recognized by the database
as functionally dependent on the columns in the grouping clause.
```

> The other supported databases will not encounter the ``OperationalError`` in the
> example above, because they will detect the functional dependency. In general,
> ``AnyValue`` is useful for working around columns in the select list that
> involve non-aggregate functions or complex expressions that are not determined
> by the database to be functionally dependent on the columns in the grouping clause.

Reviewer: ```suggestion
Other supported databases do not encounter the ``OperationalError`` in the
example above because they can detect the functional dependency. In general,
``AnyValue`` is useful when dealing with select list columns that involve
non-aggregate functions or complex expressions not recognized by the database
as functionally dependent on the columns in the grouping clause.
```
